File: research/explorations/bench-realistic/records.lua
local function run(n)
    local s = 0
    for i = 0, n - 1 do
        local v = { x = i, y = i * 2, z = i * 3 }
        local d = v.x + v.y + v.z
        local v2 = { x = v.x + 1, y = v.y, z = v.z }
        s = s + d + v2.x
    end
    return s
end

local N = tonumber(arg[1]) or 1000
-- warmup
for i = 1, 1000 do run(N) end
local iters = 10000
local clock = os.clock
local start = clock()
local r
for i = 1, iters do r = run(N) end
local elapsed = (clock() - start) * 1e9
io.write(string.format("result:     %d\n", r))
io.write(string.format("iterations: %d\n", iters))
io.write(string.format("total:      %.2fms\n", elapsed / 1e6))
io.write(string.format("per call:   %.0fns\n", elapsed / iters))

File: research/explorations/bench-realistic/records.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

type Vec3 struct{ x, y, z int64 }

func run(n int64) int64 {
	var s int64
	for i := int64(0); i < n; i++ {
		v := Vec3{i, i * 2, i * 3}
		d := v.x + v.y + v.z
		v2 := Vec3{v.x + 1, v.y, v.z}
		s += d + v2.x
	}
	return s
}

func main() {
	n := int64(1000)
	if len(os.Args) > 1 {
		if v, err := strconv.ParseInt(os.Args[1], 10, 64); err == nil {
			n = v
		}
	}
	for i := 0; i < 1000; i++ {
		run(n)
	}
	iters := int64(10000)
	start := time.Now()
	var r int64
	for i := int64(0); i < iters; i++ {
		r = run(n)
	}
	elapsed := time.Since(start).Nanoseconds()
	fmt.Printf("result:     %d\n", r)
	fmt.Printf("iterations: %d\n", iters)
	fmt.Printf("total:      %.2fms\n", float64(elapsed)/1e6)
	fmt.Printf("per call:   %dns\n", elapsed/iters)
}
